core/templates/pages/contributor-dashboard-page/modal-templates/translation-modal.controller.ts

```diff
@@ -99,34 +121,113 @@ angular.module('oppia').controller('TranslationModalController', [
+            translatedElement => (
+              translatedElement === originalElement && originalElement !== ''));
+        });
+      return states;
+    };
+
     $scope.suggestTranslatedText = function() {
-      if (!$scope.uploadingTranslation && !$scope.loadingData) {
-        SiteAnalyticsService.registerContributorDashboardSubmitSuggestionEvent(
-          'Translation');
-        $scope.uploadingTranslation = true;
-        var imagesData = ImageLocalStorageService.getStoredImagesData();
-        ImageLocalStorageService.flushStoredImagesData();
-        ContextService.resetImageSaveDestination();
-        TranslateTextService.suggestTranslatedText(
-          $scope.activeWrittenTranslation.html,
-          TranslationLanguageService.getActiveLanguageCode(),
-          imagesData, function() {
-            AlertsService.addSuccessMessage(
-              'Submitted translation for review.');
-            if ($scope.moreAvailable) {
-              var textAndAvailability = (
-                TranslateTextService.getTextToTranslate());
-              $scope.textToTranslate = textAndAvailability.text;
-              $scope.moreAvailable = textAndAvailability.more;
+      $scope.hasImgCopyError = false;
+      $scope.hasImgTextError = false;
+      $scope.originalElements = angular.element(
+        $scope.textToTranslate);
+      $scope.translatedElements = angular.element(
+        $scope.activeWrittenTranslation.html);
+
+      const foundTranslatedImageFilePaths = $scope.findImgAttributes(
+        $scope.translatedElements, 'filepath-with-value');
+      const foundTranslatedImageAltTxts = $scope.findImgAttributes(
+        $scope.translatedElements, 'alt-with-value');
+      const foundTranslatedImageDescriptions = $scope.findImgAttributes(
+        $scope.translatedElements, 'caption-with-value');
+
+      const foundOriginalImageFilePaths = $scope.findImgAttributes(
+        $scope.originalElements, 'filepath-with-value');
+      const foundOriginalImageAltTxts = $scope.findImgAttributes(
+        $scope.originalElements, 'alt-with-value');
+      const foundOriginalImageDescriptions = $scope.findImgAttributes(
+        $scope.originalElements, 'caption-with-value');
+
+      const copiedImgStates = $scope.findAvailableElements(
+        foundOriginalImageFilePaths, foundTranslatedImageFilePaths, true);
+      const duplicateImgAltTextStates = $scope.findAvailableElements(
+        foundOriginalImageAltTxts, foundTranslatedImageAltTxts, false);
+      const duplicateImgDescriptionStates = $scope.findAvailableElements(
+        foundOriginalImageDescriptions,
+        foundTranslatedImageDescriptions, false);
+
+      const hasUncopiedImgs = $scope.hasTranslationErrors(
+        copiedImgStates, false);
+      const hasDuplicateAltTexts = $scope.hasTranslationErrors(
+        duplicateImgAltTextStates, true);
+      const hasDuplicateDescriptions = $scope.hasTranslationErrors(
+        duplicateImgDescriptionStates, true);
```

Comment:
Remove all this error handling logic into helper functions. You might want to consider creating a new class to hold the error types so that you can return one error object from a call that looks like

```
const translationError = validateTranslation(angular.element(
        $scope.textToTranslate), angular.element(
        $scope.activeWrittenTranslation.html));
```

Then your checks below would look something like 

```
if (translationError.hasUncopiedImgs) {
        $scope.hasImgCopyError = true;
      }
...
```

See core/templates/pages/contributor-dashboard-page/services/translate-text.service.ts in https://github.com/oppia/oppia/pull/11757/files for an example TypeScript class.

Feel free to set up a sync if you want to discuss over meet.

Reply:
Done. Do let me know whether anything needs to be changed